MCP updates

nilspalumbo opened the pull request:
Allows `FastMCPServer` to forward image context (enabled by default) and resources (disabled by default) from the MCP server to the LLM.

Adds optional persistence between tool calls to the MCP server connection, needed by certain servers (such as Playwright) to maintain state, and adds a simple example of Wikipedia question answering with Playwright.

nilspalumbo committed:
Optional persistent connections for fastmcp client; support image content and resource forwarding

nilspalumbo committed:
Add playwright mcp example

nilspalumbo committed:
lint

nilspalumbo committed:
spacing

nilspalumbo committed:
example prompt update

nilspalumbo committed:
replace azure in example

pchalasani committed:
Add resource leak warning for unclosed persistent MCP connections

When persist_connection=True, connections created outside of context
managers may leak resources. Added __del__ method to warn developers
about unclosed connections during testing/development.

pchalasani committed:
Add comprehensive tests for MCP resource forwarding and improve container handling

- Add tests for forward_text_resources and forward_blob_resources parameters
- Add __del__ method to FastMCPClient for persistent connection cleanup warnings
- Fix blob resource MIME type preservation in file attachments
- Fix Playwright example Docker connection issues on macOS with fallback socket path
- Add container debugging and status checking to Playwright example
- Improve connection retry logic with better error reporting
- All tests pass and code meets quality standards

pchalasani commented on langroid/agent/tools/mcp/fastmcp_client.py:
```diff
@@ -209,39 +235,58 @@ async def get_tool_async(self, tool_name: str) -> Type[ToolMessage]:
 
-        if not hasattr(tool_model, "handle_async"):
-            # 3) define an arg-free handle_async() method
+        if not hasattr(tool_model, "response_async"):
```
I added the assert

pchalasani reviewed (changes requested):
looks great, added tests, fixed the example, and added tests for the image/resource/blob forwarding, persistent connection, and response_async

pchalasani committed:
example: examples/mcp/playwright-mcp.py

pchalasani committed:
feat: enhance tool handler extraction to support optional agent parameter

- Enhanced _get_tool_list() in base.py to support handle methods with optional agent parameter
- Added _analyze_handler_params() to detect parameter types via annotations or names
- Added _create_handler_wrapper() to create appropriate wrapper functions for various signatures
- Updated MCP tool generation to use single handle_async method with optional agent
- Added comprehensive tests for all handler signature variations
- Maintained backward compatibility with existing response method extraction

This allows ToolMessage classes to define handle methods with signatures:
- handle()
- handle(agent)
- handle(chat_doc)
- handle(agent, chat_doc) or handle(chat_doc, agent)

The enhancement enables MCP tools to return ChatDocument with file attachments
when an agent is provided, while falling back to simple string responses otherwise.

pchalasani commented:
@nilspalumbo I'm realizing it doesn't make sense to have both `handle_async` and `response_async` methods, since langroid will always only use the first one. So I overhauled this to only create a `handle_async` method, and also generalized `handle`, `handle_async` to accept an `agent` arg.

pchalasani committed:
tests: update test_mcp_tools

pchalasani committed:
examples/mcp/pulleteer-mcp.py

pchalasani commented:
Also added `examples/mcp/puppeteer-mcp.py` which is exactly same except uses Puppeteer MCP, and also works fine (no docker)

pchalasani committed:
Remove examples/mcp/playwright directory from PR

nilspalumbo commented:
I'm referring to the new one (the old docker image did all the setup for system dependencies, browser installation, etc.), the agent can (ignoring authentication) run the scripts for browser installation in the new example, but those do not work on my system.

Not a major issue though, it seems that the MCP server is now compatible with the current version of Chrome, so it's working without Docker when Chrome is installed.

pchalasani commented:
> it might be better to user the non-string annotations when available to better handle agent arguments which don't have `Agent` in their class names.

@nilspalumbo The problem is that the processing of the `handle_async` is not happening at run time; it's at definition time, so we can't use `instanceof`, and have to rely on strings. Or am I not understanding?

nilspalumbo commented:
This is happening at run time, so that isn't an issue; I pushed an initial version of the handler generation code which uses subclassing to a new copy of this branch.

ondonga254-code commented on examples/mcp/playwright/playwright-wikipedia.py:
```diff
@@ -84,14 +84,34 @@ async def main(question: str = "", model: str = ""):
+            f"Building Docker image {tag}... (this may take several minutes on first run)"
+        )
         path = os.path.dirname(os.path.abspath(__file__))
-        client.images.build(path=path, tag=tag)
```
Tr:
"""Builds runs the playwright-mcp Docker Image with real time log streaming and macOS specific socket handling"""

ondonga254-code commented on examples/mcp/playwright/playwright-wikipedia.py:
```diff
@@ -100,9 +120,29 @@ async def main(question: str = "", model: str = ""):
+
     # Wait for the server to start
     connected = False
-    while not connected:
```
# try connecting to the server with a cap to prevent infinite hanging if the container fails to initialize.

ondonga254-code commented on langroid/agent/tools/mcp/fastmcp_client.py:
```diff
@@ -281,9 +282,21 @@ async def call_tool_async(itself: ToolMessage) -> Any:
+
         if not hasattr(tool_model, "response_async"):
-            # 3) define an arg-free response_async() method
-            # if the tool model doesn't already have one
```
#Ensure the response includes any necessary file attachments or binary data by returning a ChatDocument with blob resources.

ondonga254-code commented on langroid/agent/tools/mcp/fastmcp_client.py:
```diff
@@ -387,7 +400,10 @@ def _convert_tool_result(
                     and self.forward_blob_resources
                 ):
                     results_file.append(
-                        FileAttachment.from_io(BytesIO(b64decode(item.resource.blob)))
```
# Ensure the attachment carries its origninal MIME type for proper client-side handling
results_file.append(
      FileAttachment.from_io(
           BytesIO(b64decode(item.resource.blob)),
            mime_type=item.resource.mimeType,

ondonga254-code commented on tests/main/test_mcp_tools.py:
```diff
@@ -9,7 +9,14 @@
     SamplingParams,
 )
 from fastmcp.client.transports import NpxStdioTransport, UvxStdioTransport
-from mcp.types import ImageContent, TextContent, Tool
```
"Updating mcp.types imports to include BlobResourceContents , EmbeddedResource , and TextResourceContents for enhanced testing of resource based interactions"

ondonga254-code commented on examples/mcp/playwright/playwright-wikipedia.py:
```diff
@@ -84,14 +84,34 @@ async def main(question: str = "", model: str = ""):
     which is shown first when sorted by year.
     """
     )
-    client = docker.from_env()
```
"Catch DockerException on from_env() to mannually point to the macOS Docker Desktop Socket at ~/.docker/run/docker.sock."

ondonga254-code reviewed (commented):
Suggested comments to fx the pending errors.
